File: sdk/typescript/tests/parity/parity.test.ts
/**
 * Cross-language parity runner (TypeScript leg).
 *
 * This test LOADS the language-agnostic fixture sdk/parity/scenario.json and
 * executes its six ordered steps against a REAL orb spawned over a UNIX domain
 * socket (the same spawn approach the contract test uses). Each step is
 * dispatched to the concrete TypeScript SDK method named in the fixture's
 * `sdk_methods.typescript` entry, and the result is asserted against the step's
 * `expected` block and skip rules.
 *
 * Static conformance (validate_sdk_spec_conformance.py) proves each step's
 * (method, path, operationId) — and now sdk_methods.typescript — resolves to a
 * real spec operation and client method. This runtime leg proves the TS SDK
 * actually drives the scenario end-to-end and produces equivalent outcomes.
 */

import * as fs from "fs";
import * as os from "os";
import * as path from "path";
import { spawn, type ChildProcess } from "child_process";
import { request as undiciRequest, Agent } from "undici";
import { OrbClient, OrbApiError } from "../../src/index";

// The fixture is the single source of truth. It lives at sdk/parity relative to
// the TS SDK dir (sdk/typescript/tests/parity → sdk/parity). Load it at runtime
// via fs so the test does not depend on tsconfig JSON-module / rootDir settings.
interface ScenarioStep {
  step: number;
  name: string;
  sdk_methods: Record<string, string>;
}
interface Scenario {
  steps: ScenarioStep[];
}
const SCENARIO_PATH = path.resolve(__dirname, "..", "..", "..", "parity", "scenario.json");
const scenario: Scenario = JSON.parse(fs.readFileSync(SCENARIO_PATH, "utf8"));

const ORB_BINARY = process.env["ORB_BINARY"] ?? "python3";
const ORB_SRC = process.env["ORB_SRC"] ?? "";
const START_TIMEOUT_MS = 45_000;

interface OrbFixture {
  client: OrbClient;
  socketPath: string;
  proc: ChildProcess;
  tmpDir: string;
  stop: () => Promise<void>;
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

async function spawnOrb(): Promise<OrbFixture> {
  const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), "orb-parity-ts-"));
  const socketPath = path.join(tmpDir, "orb.sock");
  const configPath = path.join(tmpDir, "config.json");

  const config = {
    version: "2.0.0",
    scheduler: { type: "default" },
    provider: {
      providers: [
        { name: "aws-stub", type: "aws", enabled: true, config: { region: "us-east-1" } },
      ],
    },
    storage: { type: "json" },
    server: {
      host: "127.0.0.1",
      port: 19995,
      working_dir: tmpDir,
      pid_file: path.join(tmpDir, "orb-server.pid"),
    },
    auth: { type: "none" },
    logging: { level: "ERROR" },
  };
  fs.writeFileSync(configPath, JSON.stringify(config, null, 2));

  const proc = spawn(
    ORB_BINARY,
    ["-m", "orb", "--config", configPath, "server", "start", "--foreground", "--api-only", "--socket-path", socketPath],
    {
      env: { ...process.env, PYTHONPATH: ORB_SRC, ORB_LOG_LEVEL: "ERROR" },
      stdio: ["ignore", "pipe", "pipe"],
    }
  );

  const agent = new Agent({ connect: { socketPath }, connections: 1 });
  const deadline = Date.now() + START_TIMEOUT_MS;
  let healthy = false;
  while (Date.now() < deadline) {
    await sleep(300);
    try {
      const { statusCode, body } = await undiciRequest("http://localhost/health", {
        dispatcher: agent,
        signal: AbortSignal.timeout(2000),
      });
      const chunks: Buffer[] = [];
      for await (const chunk of body) {
        chunks.push(Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk));
      }
      const json = JSON.parse(Buffer.concat(chunks).toString("utf8")) as { status: string };
      if (statusCode === 200 && (json.status === "healthy" || json.status === "degraded")) {
        healthy = true;
        break;
      }
    } catch {
      // not ready yet
    }
  }
  await agent.close();

  if (!healthy) {
    proc.kill("SIGKILL");
    fs.rmSync(tmpDir, { recursive: true, force: true });
    throw new Error(`ORB did not become healthy within ${START_TIMEOUT_MS}ms`);
  }

  const client = await OrbClient.create({
    socketPath,
    auth: { type: "none" },
    timeoutMs: 15_000,
    retry: { maxRetries: 1, baseDelayMs: 100 },
  });

  const stop = async (): Promise<void> => {
    await client.close();
    proc.kill("SIGTERM");
    await new Promise<void>((resolve) => {
      const t = setTimeout(() => { proc.kill("SIGKILL"); resolve(); }, 5000);
      proc.on("exit", () => { clearTimeout(t); resolve(); });
    });
    fs.rmSync(tmpDir, { recursive: true, force: true });
  };

  return { client, socketPath, proc, tmpDir, stop };
}

/**
 * A route-level 404/405 means the SDK is calling the wrong path/verb — always a
 * FAIL, even for conditional steps. Resource-level 404s are tolerated.
 */
function assertNotRouteLevelError(err: unknown, context: string): void {
  if (!(err instanceof OrbApiError)) return;
  if (err.statusCode === 405) {
    throw new Error(`${context}: HTTP 405 Method Not Allowed — route-level client/spec bug`);
  }
  if (err.statusCode === 404 && err.message === "Not Found" && !err.code) {
    throw new Error(`${context}: HTTP 404 with no detail — route-level missing-path bug`);
  }
}

// Variables bound across steps by the fixture's post_condition rules.
interface ParityState {
  firstTemplateId?: string;
  requestId?: string;
  machineId?: string;
}

type StepResult = "PASS" | "SKIP";

const HEALTHY_STATUSES = ["healthy", "degraded"];

describe("Parity scenario (TypeScript)", () => {
  let fixture: OrbFixture;
  const state: ParityState = {};
  const results: Record<number, StepResult> = {};

  beforeAll(async () => {
    fixture = await spawnOrb();
    console.log(`ORB started: PID=${fixture.proc.pid}, socket=${fixture.socketPath}`);
  }, START_TIMEOUT_MS + 5_000);

  afterAll(async () => {
    if (fixture) {
      await fixture.stop();
      for (const step of scenario.steps) {
        console.log(`PARITY ${step.step} ${step.name}: ${results[step.step]}`);
      }
      console.log("ORB stopped");
    }
  });

  // The fixture drives the assertions; one Jest test per fixture step.
  for (const step of scenario.steps) {
    const method = step.sdk_methods["typescript"];

    it(`step ${step.step}: ${step.name} → ${method}`, async () => {
      const c = fixture.client;
      let result: StepResult = "PASS";

      switch (step.step) {
        case 1: {
          // health_check — await client.health()
          const res = await c.health();
          expect(res.status).toBeDefined();
          expect(HEALTHY_STATUSES.includes(res.status)).toBe(true);
          break;
        }
        case 2: {
          // list_templates — await client.listTemplates()
          const res = await c.listTemplates();
          expect(res).toHaveProperty("templates");
          expect(Array.isArray(res.templates)).toBe(true);
          if (res.templates && res.templates.length > 0) {
            const first = res.templates[0] as Record<string, unknown>;
            state.firstTemplateId = (first["templateId"] ?? first["template_id"]) as string;
            console.log(`  bound firstTemplateId=${state.firstTemplateId}`);
          }
          break;
        }
        case 3: {
          // request_machines — precondition: firstTemplateId bound
          if (!state.firstTemplateId) { result = "SKIP"; break; }
          try {
            const res = await c.requestMachines({ templateId: state.firstTemplateId, count: 1 } as any);
            const requestId = ((res as any).requestId ?? (res as any).request_id) as string;
            expect(requestId).toBeTruthy();
            state.requestId = requestId;
            console.log(`  bound requestId=${state.requestId}`);
          } catch (err) {
            // A provider-level failure (no real AWS) is not a route bug.
            assertNotRouteLevelError(err, "requestMachines");
            console.log(`  requestMachines non-route error (expected without real provider): ${err}`);
            result = "SKIP";
          }
          break;
        }
        case 4: {
          // poll_request_status — precondition: requestId bound
          if (!state.requestId) { result = "SKIP"; break; }
          const res = await c.getRequestStatus(state.requestId);
          const obj = res as Record<string, unknown>;
          expect("status" in obj || "requests" in obj).toBe(true);
          const requests = obj["requests"] as Array<Record<string, unknown>> | undefined;
          if (requests && requests.length > 0) {
            const m = requests[0]["machines"] as Array<Record<string, unknown>> | undefined;
            if (m && m.length > 0) {
              state.machineId = (m[0]["machineId"] ?? m[0]["machine_id"]) as string;
            }
          }
          break;
        }
        case 5: {
          // return_machines — precondition: requestId AND a machineId
          if (!state.requestId || !state.machineId) { result = "SKIP"; break; }
          try {
            await c.returnMachines({ machineIds: [state.machineId] } as any);
          } catch (err) {
            assertNotRouteLevelError(err, "returnMachines");
            console.log(`  returnMachines non-route error (acceptable): ${err}`);
          }
          break;
        }
        case 6: {
          // list_requests — always executed
          const res = await c.listRequests();
          const obj = res as Record<string, unknown>;
          expect("requests" in obj || "data" in obj).toBe(true);
          break;
        }
        default:
          throw new Error(`unknown step ${step.step} (${step.name}) — update the TS parity runner`);
      }

      results[step.step] = result;
      console.log(`  step ${step.step} ${step.name}: ${result}`);
    });
  }
});

File: sdk/go/orb/parity_test.go
//go:build integration

package orb_test

import (
	"context"
	"encoding/json"
	"errors"
	"os"
	"path/filepath"
	"testing"
	"time"

	"github.com/finos/open-resource-broker/sdk/go/orb"
	"github.com/finos/open-resource-broker/sdk/go/testutil"
)

// ---------------------------------------------------------------------------
// Cross-language parity runner (Go leg).
//
// This test LOADS the language-agnostic fixture sdk/parity/scenario.json and
// executes its six ordered steps against a REAL orb spawned over a UDS (the
// same testutil.StartORB harness the contract tests use).  Each step is
// dispatched to the concrete Go SDK method named in the fixture's
// sdk_methods.go entry, and the result is asserted against the step's
// `expected` block and skip rules.
//
// Static conformance (validate_sdk_spec_conformance.py) proves each step's
// (method, path, operationId) — and now sdk_methods.go — resolves to a real
// spec operation and client method.  This runtime leg proves the Go SDK
// actually drives the scenario end-to-end and produces equivalent outcomes.
// ---------------------------------------------------------------------------

type parityStep struct {
	Step        int               `json:"step"`
	Name        string            `json:"name"`
	SDKMethods  map[string]string `json:"sdk_methods"`
	Expected    parityExpected    `json:"expected"`
	Precond     string            `json:"precondition"`
	PostCond    *parityPostCond   `json:"post_condition"`
}

type parityExpected struct {
	// http_status may be a single int or a list of ints in the fixture.
	HTTPStatus  json.RawMessage `json:"http_status"`
	StatusShape *struct {
		StatusMustBeOneOf []string `json:"status_must_be_one_of"`
	} `json:"status_shape"`
}

type parityPostCond struct {
	Bind    string `json:"bind"`
	IfEmpty string `json:"if_empty"`
}

type parityScenario struct {
	Steps []parityStep `json:"steps"`
}

// parityState carries the variables bound across steps.
type parityState struct {
	firstTemplateID string
	requestID       string
	machineID       string
}

// stepResult is the PASS/SKIP outcome recorded per step (FAIL is a t.Error).
type stepResult string

const (
	stepPass stepResult = "PASS"
	stepSkip stepResult = "SKIP"
)

func loadScenario(t *testing.T) parityScenario {
	t.Helper()
	path := filepath.Join("..", "..", "parity", "scenario.json")
	raw, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("parity: reading %s: %v", path, err)
	}
	var s parityScenario
	if err := json.Unmarshal(raw, &s); err != nil {
		t.Fatalf("parity: parsing scenario.json: %v", err)
	}
	if len(s.Steps) == 0 {
		t.Fatal("parity: scenario.json declares no steps")
	}
	return s
}

// assertNotRouteLevel fails the step if err is a route-level 404/405 — those
// mean the Go SDK is calling the wrong path/verb, which must always be FAIL
// even for conditional steps.  Resource-level 404s are tolerated by callers.
func assertNotRouteLevel(t *testing.T, step string, err error) {
	t.Helper()
	if err == nil {
		return
	}
	var apiErr *orb.OrbApiError
	if errors.As(err, &apiErr) {
		if apiErr.StatusCode == 405 {
			t.Errorf("parity[%s]: FAIL — HTTP 405 Method Not Allowed (wrong verb, route-level bug)", step)
		}
		if apiErr.StatusCode == 404 && apiErr.Code == "" && apiErr.Message == "Not Found" {
			t.Errorf("parity[%s]: FAIL — HTTP 404 with no detail (route-level missing path bug)", step)
		}
	}
}

func TestParityScenario(t *testing.T) {
	scenario := loadScenario(t)
	c := testutil.StartORB(t)
	ctx, cancel := context.WithTimeout(context.Background(), 90*time.Second)
	defer cancel()

	state := &parityState{}
	results := make(map[int]stepResult)

	for _, step := range scenario.Steps {
		method := step.SDKMethods["go"]
		if method == "" {
			t.Fatalf("parity[step %d %s]: fixture has no sdk_methods.go entry", step.Step, step.Name)
		}
		t.Logf("step %d %-22s → %s", step.Step, step.Name, method)

		result := runParityStep(t, ctx, c, step, state)
		results[step.Step] = result
		t.Logf("  step %d %s: %s", step.Step, step.Name, result)
	}

	// Summary line — mirrors the README's "each step is PASS, SKIP, or FAIL".
	for _, step := range scenario.Steps {
		t.Logf("PARITY %d %-22s %s", step.Step, step.Name, results[step.Step])
	}
}

// runParityStep dispatches one step to its concrete Go SDK method and asserts
// the fixture's expectations.  Preconditions that are not met produce a SKIP.
func runParityStep(
	t *testing.T,
	ctx context.Context,
	c *orb.Client,
	step parityStep,
	state *parityState,
) stepResult {
	t.Helper()
	tag := step.Name

	switch step.Step {
	case 1: // health_check — client.Health(ctx)
		resp, err := c.Health(ctx)
		assertNotRouteLevel(t, tag, err)
		if err != nil {
			t.Errorf("parity[%s]: FAIL — Health returned error: %v", tag, err)
			return stepPass
		}
		if !oneOfExpectedStatus(step, resp.Status) {
			t.Errorf("parity[%s]: FAIL — health status %q not in expected set", tag, resp.Status)
		}
		return stepPass

	case 2: // list_templates — client.ListTemplates(ctx)
		templates, err := c.ListTemplates(ctx)
		assertNotRouteLevel(t, tag, err)
		if err != nil {
			t.Errorf("parity[%s]: FAIL — ListTemplates returned error: %v", tag, err)
			return stepPass
		}
		if len(templates) > 0 {
			state.firstTemplateID = templates[0].TemplateID
			t.Logf("  bound first_template_id=%s", state.firstTemplateID)
		} else {
			t.Logf("  no templates — steps 3-5 will skip per fixture if_empty rule")
		}
		return stepPass

	case 3: // request_machines — precondition: first_template_id bound
		if state.firstTemplateID == "" {
			return stepSkip
		}
		resp, err := c.RequestMachines(ctx, orb.RequestMachinesRequest{
			TemplateID: state.firstTemplateID,
			Count:      1,
		})
		assertNotRouteLevel(t, tag, err)
		if err != nil {
			// A provider-level failure (no real AWS) is not a route bug; the
			// route was reachable.  Treat as skip of the downstream chain.
			t.Logf("  RequestMachines returned non-route error (expected without real provider): %v", err)
			return stepSkip
		}
		if resp.RequestID == "" {
			t.Errorf("parity[%s]: FAIL — 2xx but empty request_id binding", tag)
		}
		state.requestID = resp.RequestID
		t.Logf("  bound request_id=%s", state.requestID)
		return stepPass

	case 4: // poll_request_status — precondition: request_id bound
		if state.requestID == "" {
			return stepSkip
		}
		status, err := c.GetRequestStatus(ctx, state.requestID, false)
		assertNotRouteLevel(t, tag, err)
		if err != nil {
			t.Errorf("parity[%s]: FAIL — GetRequestStatus returned error: %v", tag, err)
			return stepPass
		}
		if status.Status == "" {
			t.Errorf("parity[%s]: FAIL — status field empty", tag)
		}
		for _, m := range status.Machines {
			if m.MachineID != "" {
				state.machineID = m.MachineID
				break
			}
		}
		return stepPass

	case 5: // return_machines — precondition: request_id AND a machine_id
		if state.requestID == "" || state.machineID == "" {
			return stepSkip
		}
		err := c.ReturnMachines(ctx, []string{state.machineID})
		assertNotRouteLevel(t, tag, err)
		if err != nil {
			t.Logf("  ReturnMachines returned non-route error (acceptable): %v", err)
		}
		return stepPass

	case 6: // list_requests — always executed
		_, err := c.ListRequests(ctx)
		assertNotRouteLevel(t, tag, err)
		if err != nil {
			t.Errorf("parity[%s]: FAIL — ListRequests returned error: %v", tag, err)
		}
		return stepPass

	default:
		t.Fatalf("parity: unknown step number %d (%s) — update the Go parity runner", step.Step, step.Name)
		return stepPass
	}
}

// oneOfExpectedStatus checks a health status string against the fixture's
// status_shape.status_must_be_one_of list (falling back to the canonical set).
func oneOfExpectedStatus(step parityStep, status string) bool {
	allowed := []string{"healthy", "degraded"}
	if step.Expected.StatusShape != nil && len(step.Expected.StatusShape.StatusMustBeOneOf) > 0 {
		allowed = step.Expected.StatusShape.StatusMustBeOneOf
	}
	for _, a := range allowed {
		if status == a {
			return true
		}
	}
	return false
}
